src/Raven.Server/Documents/AI/Embeddings/EmbeddingsHelper.cs

```diff
@@ -0,0 +1,114 @@
+    /// The system uses these values to generate test embeddings and validate that the
+    /// language model properly responds to embedding generation requests.
+    /// </remarks>
+    internal static readonly List<string> ValuesListToVerifyConnection = ["TestValue", "TestValue2"];
+    private static readonly IBlake2B Hash;
+
+    static EmbeddingsHelper()
+    {
+        Hash = Blake2BFactory.Instance.Create(new Blake2BConfig
+        {
+            HashSizeInBits = 256
+        });
+    }
+
+    public static string CalculateInputValueHash(string value)
+    {
+        return Hash.ComputeHash(value).AsHexString(uppercase: true);
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public static string GenerateDestinationAttachmentName(in string prefix, in string originalAttachmentName, in VectorEmbeddingType quantization)
+    {
+        var suffix = quantization switch
+        {
+            VectorEmbeddingType.Int8 => "_int8",
+            VectorEmbeddingType.Binary => "_binary",
+            _ => string.Empty
+        };
+        
+        return $"{prefix}{originalAttachmentName}{suffix}";
+    }
+    
+    public static string GetEmbeddingDocumentId(string documentId)
+    {
+        return $"{documentId}/embeddings";
+    }
+
+    public static string GetEmbeddingDocumentCollectionName(string sourceCollectionName)
+    {
+        return $"{sourceCollectionName}/embeddings";
+    }
+
+    public static string GetPrefixForAttachmentInEmbeddingsDocument(EmbeddingsGenerationTaskIdentifier embeddingsGenerationTaskIdentifier, string path)
+    {
+        return $"{embeddingsGenerationTaskIdentifier.Value}_{path}_";
+    }
+
+    public static string GetEmbeddingCacheDocumentId(AiConnectionStringIdentifier aiConnectionStringIdentifier, string valueHash, in VectorEmbeddingType targetQuantization)
+    {
+        var suffix = targetQuantization switch
+        {
+            Raven.Client.Documents.Indexes.Vector.VectorEmbeddingType.Single => string.Empty,
+            Raven.Client.Documents.Indexes.Vector.VectorEmbeddingType.Int8 => "/int8",
+            Raven.Client.Documents.Indexes.Vector.VectorEmbeddingType.Binary => "/binary",
+            _ => throw new ArgumentException($"Unknown quantization type '{targetQuantization}'")
+        };
+        
+        return $"embeddings-cache/{aiConnectionStringIdentifier.Value}/{valueHash}{suffix}";
+    }
+
+    public static EmbeddingValue CreateEmbeddingValue(ReadOnlyMemory<float> embedding, VectorEmbeddingType quantization)
+    {
+        switch (quantization)
+        {
+            case VectorEmbeddingType.Single:
+                return new EmbeddingValue(embedding, embedding.Length * sizeof(float));
+            case VectorEmbeddingType.Int8:
+            {   var dest = MemoryMarshal.Cast<float, sbyte>(embedding.Span);
+                if (VectorQuantizer.TryToInt8(embedding.Span, dest, out int usedBytes) == false)
```

Comment:
You are trying to do an in-place quantization - are you sure that the source data isn't changed if this fails?

Reply:
Yes, this is our implementation.